go/proxyd/backend.go

```diff
@@ -385,30 +385,60 @@ func (b *Backend) doForward(ctx context.Context, rpcReq *RPCReq) (*RPCRes, error
 		return nil, wrapErr(err, "error reading response body")
 	}
 
-	res := new(RPCRes)
-	if err := json.Unmarshal(resB, res); err != nil {
+	var res []*RPCRes
+	if err := json.Unmarshal(resB, &res); err != nil {
+		return nil, ErrBackendBadResponse
+	}
+
+	// Alas! Certain node providers (Infura) always return a single JSON object for some types of errors
+	if len(rpcReqs) != len(res) {
 		return nil, ErrBackendBadResponse
 	}
 
 	// capture the HTTP status code in the response. this will only
 	// ever be 400 given the status check on line 318 above.
 	if httpRes.StatusCode != 200 {
-		res.Error.HTTPErrorCode = httpRes.StatusCode
+		for _, res := range res {
+			res.Error.HTTPErrorCode = httpRes.StatusCode
+		}
 	}
 
+	sortBatchRPCResponse(rpcReqs, res)
 	return res, nil
 }
 
+// sortBatchRPCResponse sorts the RPCRes slice according to the position of its corresponding ID in the RPCReq slice
+func sortBatchRPCResponse(req []*RPCReq, res []*RPCRes) {
+	pos := make(map[string]int, len(req))
+	for i, r := range req {
+		key := string(r.ID)
+		if _, ok := pos[key]; ok {
+			panic("bug! detected requests with duplicate IDs")
```

Comment:
We shouldn't `panic` on user input like this. Why do we need to sort the response at all?

Reply:
This should only happen if we're not sanitizing forwarded requests correctly. We ensure that a forwarded batch request does not contain duplicate ids at [Server.handleBatchRPC](https://github.com/ethereum-optimism/optimism/pull/2480/files/19a5563c000361b7b1130da7bc72edf54bcdbd24#diff-3b422a2479068b7d149561a3ea8cdd66abba269afa3c400c96230bd9d7f78183R214-R219). Since it'll be a bug otherwise, and not bad user input, a panic seems appropriate.

According to the JSON-RPC spec, response objects in a batch may be returned in any order. Which is OK because the Request ID fields can be used to reference a response. However, to simplify the implementation in server.go, `Backend.Forward` returns responses in the same order as the request. Sorting responses achieves this.

AFAICT, both Alchemy and Infura don't return responses out of order. But there's no guarantee this will always be the case as not doing so doesn't technically violate the JSON-RPC spec.